src/vs/gitpod/node/gitpodInsightsAppender.ts

```diff
@@ -0,0 +1,181 @@
+	getTokenRequest.setHost(gitpodApiHost);
+	getTokenRequest.addScope('function:trackEvent');
+
+	const getTokenResponse = await util.promisify(supervisor.token.getToken.bind(supervisor.token, getTokenRequest, supervisor.metadata, {
+		deadline: Date.now() + supervisor.deadlines.long
+	}))();
+	const serverToken = getTokenResponse.getToken();
+
+	return {
+		serverToken,
+		gitpodHost,
+		gitpodApiEndpoint,
+		workspaceId,
+		instanceId
+	};
+}
+
+async function getClient(serverToken: string, gitpodHost: string, gitpodApiEndpoint: string): Promise<GitpodConnection> {
+	const factory = new JsonRpcProxyFactory<GitpodServer>();
+	const gitpodService = new GitpodServiceImpl<GitpodClient, GitpodServer>(factory.createProxy()) as GitpodConnection;
+
+	const webSocket = new (ReconnectingWebSocket as any)(gitpodApiEndpoint, undefined, {
+		maxReconnectionDelay: 10000,
+		minReconnectionDelay: 1000,
+		reconnectionDelayGrowFactor: 1.3,
+		connectionTimeout: 10000,
+		maxRetries: Infinity,
+		debug: false,
+		startClosed: false,
+		WebSocket: class extends WebSocket {
+			constructor(address: string, protocols?: string | string[]) {
+				super(address, protocols, {
+					headers: {
+						'Origin': new URL(gitpodHost).origin,
+						'Authorization': `Bearer ${serverToken}`
+					}
+				});
+			}
+		}
+	});
+	webSocket.onerror = console.error;
+	doListen({
+		webSocket: webSocket as any,
+		onConnection: connection => factory.listen(connection),
+		logger: new ConsoleLogger()
+	});
+
+	return gitpodService;
+}
+
+const formatEventName = (str: string) => {
+	return str
+		.replace(/^[A-Z]/g, letter => letter.toLowerCase())
+		.replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`)
+		.replace(/[^\w]/g, '_');
+};
+
+export class GitpodInsightsAppender implements ITelemetryAppender {
+
+	private _asyncAIClient: Promise<GitpodConnection> | null;
+	private _defaultData: { [key: string]: any } = Object.create(null);
+
+	constructor() {
+		this._asyncAIClient = null;
+	}
+
+	private _withAIClient(callback: (aiClient: Pick<GitpodServer, 'trackEvent'>) => void): void {
+		if (!this._asyncAIClient) {
+			this._asyncAIClient = getSupervisorData().then(
+				(supervisorData) => {
+					this._defaultData['workspaceId'] = supervisorData.workspaceId;
+					this._defaultData['instanceId'] = supervisorData.instanceId;
+
+					return getClient(supervisorData.serverToken, supervisorData.gitpodHost, supervisorData.gitpodApiEndpoint);
+				}
+			);
+		}
+
+		this._asyncAIClient.then(
+			(aiClient) => {
+				callback(aiClient.server);
+			},
+			(err) => {
+				onUnexpectedError(err);
+				console.error(err);
+			}
+		);
+	}
+
+	log(eventName: string, data?: any): void {
+		data = mixin(data, this._defaultData);
+		data = validateTelemetryData(data);
+
+		if (eventName !== 'workbenchActionExecuted') {
+			return;
+		}
+
+		this._withAIClient((aiClient) => aiClient.trackEvent({
+			event: `vscode_${formatEventName(eventName)}`,
+			properties: data.properties
```

Comment:
I think workspaceid and workspaceinstanceid shokld be added to properties here.

We probably should prefix them somehow to avoid collisions